### DISCUSSION TITLE
originalError: Connection refused error

### DISCUSSION THREAD

**Discussion by nateluk** [pocketbase/pocketbase#1034] [Q&A] [answered]
Hi, i'm trying to use pocketbase for a basic flutter app but upon trying to connect to my db, I'm getting the following error:
`Unhandled Exception: ClientException: {url: http://127.0.0.1:8090/api/collections/notes/records?page=1&perPage=30, isAbort: true, statusCode: 0, response: {}, originalError: Connection refused}`

if I open the url above it correctly displays the data in JSON format in my browser
My dart code to generate this error is simple running:
```    
var db = PocketBase('http://127.0.0.1:8090');
var data = await db.records.getList('notes');
```

I've made sure that the API rules for my collection are unlocked (public)

Is there something I'm missing? 

Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
How are you running your flutter app?

If you are running for example using Android emulator then you can try changing `127.0.0.1` with `10.0.2.2` (the loopback that avd uses for the host localhost) or alternatively with the IP of you local network (to find it you can run `ifconfig` on linux)

**Reply by nateluk** [reply]
thanks for your response. How would I call `./pocketbase serve` ? Would I need to change that call?

**Reply by ganigeorgiev** [reply]
If you are running the flutter app inside an android emulator they have their own localhost. A more info for the internal avd network space you could find in https://developer.android.com/studio/run/emulator-networking.

Usually it shouldn't be necessary to change the `pocketbase serve` startup address during development but depending on your network settings and the os you are running, you may have to change it to match with the ip of your local network or at least to bind to all local ipv4 interfaces using `0.0.0.0` (or `::` for ipv6). If necessary, to change the PocketBase address you can use the `--http` flag:
```
./pocketbase serve --http="0.0.0.0:8090"
```

Alternatively, you can also try something like [ngrok](https://ngrok.com/) and use the generated address in your flutter app (this will also allow you to access your local development env from your real mobile device even if you are on a different network)

**Reply by nateluk** [reply]
I tried using `10.0.0.2:8090` but am receiving this error when trying to serve `listen tcp 10.0.2.2:8090: bind: can't assign requested address` 

Any ideas to the issue?

I'm using linux machine if that helps

**Reply by ganigeorgiev** [reply]
I think there is some confusion.

The IP `10.0.2.2`  should be used as a replacement of `127.0.0.1` in your flutter app because as mentioned above, if you are running inside an emulator they have their own localhost and the 10.0.2.2 targets the localhost of your development machine, where PocketBase is running.

The PocketBase executable needs to bind to your localhost (aka. `127.0.0.1`) in order the above IP to work.

**Reply by nateluk** [reply]
thank you - that solved the problem! sorry for the confusion and thank you again for taking the time to suggest solutions

**Reply by lbarria77** [reply]
Hello how are you, in my case I am emulating on a physical device through adb via wifi, there is no way to make the connection I always get a connection refused or time out error. I have already tried in all possible ways with http://localhost, http://127.0.0.1 and with my pc ip. If you could help me please I would appreciate it. Greeting

**Reply by ganigeorgiev** [reply]
@lbarria77 I'm not sure that I understand the described issue. If you are using an emulator, in order to access the host machine localhost you can check comment https://github.com/pocketbase/pocketbase/discussions/1034#discussioncomment-4153707.

Alternatively, you can try "tunnel" your PocketBase app through [ngrok](https://ngrok.com/) or similar  (it will create a special public address that you should be able to request from your device).

**Reply by lbarria77** [reply]
ok what happens is that I use my physical device to emulate, but it doesn't connect pocketbase gives me this error - Unhandled Exception: ClientException: {url: http://127.0.0.1:8090/api/collections/notes/records? page=1&perPage=30, isAbort: true, statusCode: 0, response: {}, originalError: Connection refused} - I already tried changing 127.0.0.1 for 10.0.2.2, for localhost and for my pc's ip and nothing. I had to use the android studio emulator and change it to 10.0.2.2 and that's how it works but what I use is my physical device and there is no way to connect it. greeting

**Reply by ganigeorgiev** [reply]
If both your android and desktop device are connected to the same wifi network, then you should be able to use the local IP assigned to your desktop (you may need to start the PocketBase executable with `./pocketbase serve --http="0.0.0.0:8090"`). If the above doesn't work, then I'm not sure. There are way too many reasons why this may not work depending on your local network setup, eg. it could be restriction by an antivirus software, firewall, etc.

The easiest workaround would be to use [ngrok](https://ngrok.com/) (the devices even don't have to be in the same network).

**Reply by lbarria77** [reply]
Thank you very much for answering, I appreciate it

**Reply by lbarria77** [reply]
hello what I did was run the server like this final db = PocketBase('http://my-ip:8090'); and then that ip was assigned to my pocketbase and it worked on my physical device. Thank you very much
